File: C++/BinaryTreeZigzagLevelOrderTraversal.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
	vector<vector<int> > ret;
	vector<vector<int> > zigzagLevelOrder(TreeNode *root) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function
		ret.clear();
		if(!root)
			return ret;
		queue<TreeNode*> q;
		q.push(root);
		int pre = 1;
		int next = 0;
		bool direct = false;
		vector<int> list;
		while(!q.empty())
		{
			TreeNode* cur = q.front();
			q.pop();
			list.push_back(cur->val);
			pre--;
			if(cur->left)
			{
				q.push(cur->left);
				next++;
			}
			if(cur->right)
			{
				q.push(cur->right);
				next++;
			}

			if(pre == 0)
			{
				pre = next;
				next = 0;
				if(direct)
					reverse(list.begin(), list.end());
				ret.push_back(list);
				list.clear();
				direct = !direct;
			}
		}
		return ret;
	}
};

File: JAVA/BinaryTreeZigzagLevelOrderTraversal.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Queue;

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;

	TreeNode(int x) {
		val = x;
	}
}

class Solution {
	ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();

	public ArrayList<ArrayList<Integer>> zigzagLevelOrder(TreeNode root) {
		// Start typing your Java solution below
		// DO NOT write main() function
		ret.clear();
		if (root == null)
			return ret;
		boolean tag = false;
		Queue<TreeNode> q = new LinkedList<TreeNode>();
		int curLevel = 1;
		int nextLevel = 0;
		q.add(root);
		ArrayList<Integer> list = new ArrayList<Integer>();
		while (!q.isEmpty()) {
			curLevel--;
			TreeNode node = q.poll();
			list.add(node.val);
			if (node.left != null) {
				q.add(node.left);
				nextLevel++;
			}
			if (node.right != null) {
				q.add(node.right);
				nextLevel++;
			}

			if (curLevel == 0) {
				if(tag)
					Collections.reverse(list);
				ret.add(new ArrayList<Integer>(list));
				list.clear();
				curLevel = nextLevel;
				nextLevel = 0;
				tag = !tag;
			}
		}
		return ret;
	}
}

public class BinaryTreeZigzagLevelOrderTraversal {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		TreeNode root = new TreeNode(1);
		TreeNode n2 = new TreeNode(2);
		TreeNode n3 = new TreeNode(3);
		TreeNode n4 = new TreeNode(4);
		TreeNode n5 = new TreeNode(5);
		root.left = n2;
		root.right = n3;
	}

}
